luarules/gadgets/unit_reactive_armor.lua

```diff
@@ -0,0 +1,378 @@
+	if armoredUnitDefs[unitDefID] then
+		local defData = armoredUnitDefs[unitDefID]
+		if not defData.first or checkReactiveArmor(unitID, unitDefID, defData) then
+			unitArmorHealth[unitID] = defData.health
+			spSetUnitRulesParam(unitID, "reactiveArmorHealth", defData.health)
+		end
+	end
+end
+
+function gadget:UnitDestroyed(unitID, unitDefID, unitTeam)
+	unitArmorFrames[unitID] = nil
+	unitArmorHealth[unitID] = nil
+	regenerateFrame[unitID] = nil
+end
+
+function gadget:UnitDamaged(unitID, unitDefID, unitTeam, damage, paralyzer)
+	if armoredUnitDefs[unitDefID] and not paralyzer and damage > 0 then
+		doArmorDamage(unitID, unitDefID, damage)
+		regenerateFrame[unitID] = combatEndFrame
+	end
+end
+
+-- Lifecycle
+
+---Damage (or repair!) a unit's reactive armor directly, without damaging the unit.
+GG.AddReactiveArmorDamage = function(unitID, damage)
+	return doArmorDamage(unitID, spGetUnitDefID(unitID), damage)
+end
+
+---Get the current armor health remaining of a unit with reactive armor.
+GG.GetReactiveArmorHealth = function(unitID)
+	return unitArmorHealth[unitID]
+end
+
+function gadget:Initialize()
+	if not next(armoredUnitDefs) then
+		gadgetHandler:RemoveGadget()
+		return
+	end
+
+	callFromLus = Spring.UnitScript.CallAsUnit
+	gameFrame = Spring.GetGameFrame()
+	combatEndFrame = gameFrame + unitCombatDuration
+
+	if gameFrame <= 0 then
+		return
+	end
+
+	-- We handle `/luarules reload` mostly correctly except for out-of-combat timers.
+
+	-- See getArmorRestoreFrames. This might be its poor copy.
+	local function resetArmorRestoreFrames(defData, progress)
+		local pieces = defData.pieces
+		local frames = defData.frames
+		if pieces == 1 then
+			return { countdown = frames, [0] = true }
+		else
+			local restoreFrames = table.new(0, pieces + 1)
+			restoreFrames.countdown = frames
+			for piece = 1, pieces do
+				local framesRemaining = gameSpeed * (piece - 1)
+				if framesRemaining <= frames then
+					restoreFrames[framesRemaining] = piece
+				end
+			end
+			return restoreFrames
+		end
+	end
+
+	for _, unitID in ipairs(Spring.GetAllUnits()) do
+		if not Spring.GetUnitIsBeingBuilt(unitID) then
+			local defData = armoredUnitDefs[spGetUnitDefID(unitID)]
+
+			if defData then
+				local armorHealth = Spring.GetUnitRulesParam(unitID, "reactiveArmorHealth")
+
+				gadget:UnitFinished(unitID, spGetUnitDefID(unitID)) -- replaces rules params
+
+				if armorHealth and armorHealth < defData.health then
+					unitArmorFrames[unitID] = nil -- we forget our next regeneration frame :c
+					unitArmorHealth[unitID] = armorHealth
+					spSetUnitRulesParam(unitID, "reactiveArmorHealth", armorHealth)
+				else
+					unitArmorFrames[unitID] = resetArmorRestoreFrames(defData)
+					unitArmorHealth[unitID] = nil
+					spSetUnitRulesParam(unitID, "reactiveArmorHealth", false)
+				end
+			end
+		end
+	end
+end
+
+function gadget:Shutdown()
+	GG.AddReactiveArmorDamage = function() return false end
+	GG.GetReactiveArmorHealth = function() end
```

Comment:
If these are in Shutdown then the init should probably also be in Initialize. else you get the "real" functions if the gadget is left in an uninitialized state due to `enabled = false`